From fb07152b0e2065b9c5607eaa6013a8d31a30a481 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Tue, 8 Apr 2025 12:04:24 +0100
Subject: [PATCH] ldconfig: Add ldconfig.trigger no argument execution mode

Typically package management invocation passes in additional arguments
(e.g. "postinst" "postinstall" "post-commit") to mainstainer scripts,
hooks, scriplets, triggers. And to invoke ldconfig without any
arguments, one typically needs to create a shell script calls ldconfig
without arguments.

With this code change, one instead can symlink ldconfig as
ldconfig.trigger into package management hooks.d directories, or use
it as the scriplet shebang #!/sbin/ldconfig.trigger, which would
effectively call ldconfig without any arguments.

This would enable to update ldconfig caches, without having a shell
installed. This is primarily aimed at distro-less container
installations which often lack or remove shell.

Signed-off-by: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
---
 elf/ldconfig.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/elf/ldconfig.c b/elf/ldconfig.c
index 39b154c72c..3a31ef77d2 100644
--- a/elf/ldconfig.c
+++ b/elf/ldconfig.c
@@ -1182,7 +1182,12 @@ main (int argc, char **argv)
 
   /* Parse and process arguments.  */
   int remaining;
-  argp_parse (&argp, argc, argv, 0, &remaining, NULL);
+  if (strcmp (program_invocation_short_name, "ldconfig.trigger") != 0)
+    /* Only parse arguments, when not called as a package trigger */
+    argp_parse (&argp, argc, argv, 0, &remaining, NULL);
+  else
+    /* Discard all remaining arguments */
+    remaining = argc;
 
   /* Remaining arguments are additional directories if opt_manual_link
      is not set.  */
-- 
2.43.0

